### ISSUE TITLE
SubscribeToUserDataUpdatesAsync not working in 8.0

### ISSUE THREAD

**Issue by HypsyNZ** [JKorf/Binance.Net#1000]
I just migrated from 7 => 8 in some test branches

`(DataEvent<BinanceStreamOrderUpdate> data)` is always `null`

`data.Data.Event` is never `executionReport`

Before I migrated everything was working perfectly.

It seems like subscriptions for User Streams aren't working at all, however other forms of socket are working (ticker updates etc)

Not sure which package has the issue right now.

https://github.com/HypsyNZ/CryptoExchange.Net/tree/binance.net-8.0-test
https://github.com/HypsyNZ/Binance.Net/tree/binance.net-8.0-test
https://github.com/HypsyNZ/BinanceTrader.NET/tree/binance.net-8.0-test

**Comment by JKorf** [comment]
Hi, It's working for me locally. Could you share a bit more info? 
Can you successfully subscribe with the listenKey?
Do you sent a KeepAliveUserStreamAsync every +- 30 minutes?
Can you turn on Trace logging to see what is received on the socket?

**Comment by HypsyNZ** [comment]
> Hi, It's working for me locally. Could you share a bit more info? 

> Can you successfully subscribe with the listenKey? 

Yes.

Do you sent a KeepAliveUserStreamAsync every +- 30 minutes? 

Yes.

>Can you turn on Trace logging to see what is received on the socket?

There is no output I assuming that means the subscription isn't working at all.

----

How are you testing this? My real world order updates aren't received with v8 but switching back to v7 everything works again.

**Comment by JKorf** [comment]
I have it working on an external application and works as well on a local test. Is it specifically margin? I tested spot

**Comment by HypsyNZ** [comment]
> I have it working on an external application and works as well on a local test. Is it specifically margin? I tested spot

That shouldn't matter but yes it is margin

**Comment by JKorf** [comment]
Also works fine for me. You said you don't have any log output which seems weird to me.. Did you change the log level to trace in the client orders?
`LogLevel = LogLevel.Trace`
There should be output.

**Comment by HypsyNZ** [comment]
Okay.

I gave you everything you needed to help me if you actually wanted to.

**Comment by HypsyNZ** [comment]
Also the issue is 100% in one of your repos.

**Comment by HypsyNZ** [comment]
Re-opening so other people can see this.

**Comment by JKorf** [comment]
Excuse me? I'm trying to help you on a sunday morning on a project I'm providing to you open source without any charge. I don't think you have any right to act this entitled. 

Anyway, I'm not saying it's not in my repo, I'm just trying to pinpoint the issue. Which I'm asking you info for because I can't reproduce it. Doesn't seem strange to me.. If you don't want to provide this, just stay on 7.2.8. The beauty of open source is that you don't need to update, you can just take the source code.

**Comment by HypsyNZ** [comment]
It took me time and effort to prepare those test branches to report the issue and you are pretending it doesn't exist.

**Comment by arkoala** [comment]
Excuse me, similar issue here with SubscribeToKlineUpdatesAsync
Working fine on 7.2.7, not working on 8.0.0

From a humble position, I must say: Thank you in advance for your great support and maintenance

Task return Success = true, but after that, Action is never called.
Working on Win10Pro with WPF project (8.0.0)
The one that works is a Win10Pro with UWP project (7.2.7)
I do not think project type affects, but just in case...

**Comment by JKorf** [comment]
Thanks for the info, I'll look into it. Might be an issue regarding a different execution environment, let me check

**Comment by JKorf** [comment]
@arkoala Still not able to reproduce it sadly. Could you provide some info?
Trace logging of the socket connection would help, you can enable trace logging by setting it in the client options:
````csharp

var client = new BinanceSocketClient(new BinanceSocketClientOptions{
  LogLevel = LogLevel.Trace
});

````

For me it just outputs the data:
```
2022/02/20 13:05:08:566 | Debug | Binance    | Socket 1 new socket created for wss://stream.binance.com:9443/stream
2022/02/20 13:05:08:584 | Debug | Binance    | Socket 1 connecting
2022/02/20 13:05:10:020 | Trace | Binance    | Socket 1 connection succeeded, starting communication
2022/02/20 13:05:10:080 | Debug | Binance    | Socket 1 connected
2022/02/20 13:05:10:133 | Debug | Binance    | Socket 1 connected to wss://stream.binance.com:9443/stream with request {"method":"SUBSCRIBE","params":["ethusdt@kline_1d"],"id":1}
2022/02/20 13:05:10:137 | Debug | Binance    | Socket 1 sending data: {"method":"SUBSCRIBE","params":["ethusdt@kline_1d"],"id":1}
2022/02/20 13:05:10:138 | Trace | Binance    | Socket 1 Adding 59 to sent buffer
2022/02/20 13:05:10:144 | Trace | Binance    | Socket 1 sent 59 bytes
2022/02/20 13:05:10:421 | Trace | Binance    | Socket 1 received 22 bytes in single message
2022/02/20 13:05:10:424 | Trace | Binance    | Socket 1 received data: {"result":null,"id":1}
2022/02/20 13:05:10:462 | Trace | Binance    | Socket 1 Subscription completed
2022/02/20 13:05:11:211 | Trace | Binance    | Socket 1 received 377 bytes in single message
2022/02/20 13:05:11:229 | Trace | Binance    | Socket 1 received data: {"stream":"ethusdt@kline_1d","data":{"e":"kline","E":1645358708907,"s":"ETHUSDT","k":{"t":1645315200000,"T":1645401599999,"s":"ETHUSDT","i":"1d","f":761389716,"L":761706898,"o":"2763.63000000","c":"2641.23000000","h":"2764.64000000","l":"2600.00000000","v":"210379.89010000","n":317183,"x":false,"q":"559673763.13692000","V":"95866.07850000","Q":"254837451.01154000","B":"0"}}}
2022/02/20 13:05:11:379 | Trace | Binance    | Socket 1 message processed in 92ms
2022/02/20 13:05:13:343 | Trace | Binance    | Socket 1 received 377 bytes in single message
2022/02/20 13:05:13:344 | Trace | Binance    | Socket 1 received data: {"stream":"ethusdt@kline_1d","data":{"e":"kline","E":1645358711041,"s":"ETHUSDT","k":{"t":1645315200000,"T":1645401599999,"s":"ETHUSDT","i":"1d","f":761389716,"L":761706901,"o":"2763.63000000","c":"2641.22000000","h":"2764.64000000","l":"2600.00000000","v":"210380.29070000","n":317186,"x":false,"q":"559674821.20940800","V":"95866.25410000","Q":"254837914.81152800","B":"0"}}}
```

Are you doing any other things with the library while subscribing/being subscribed or is this the only operation at the time?

**Comment by arkoala** [comment]
Hi, 

I just activated the logLevel.
Providing two sets of logs. It worked for just one event for the first time in the first set of logs. No more events after that.
Test done with: 
var result = await socketClient.SpotStreams.SubscribeToKlineUpdatesAsync ("BTCBUSD", KlineInterval.FifteenMinutes, data =>
                    {
                        OnSubscribedSymbolMessage(data.Data);
});


Excuse my Spanish environment on some messages. If you need translations, just ask for it.

El subproceso 0x1cc4 terminó con código 0 (0x0).
2022/02/20 13:42:34:162 | Debug | Binance    | Socket 1 connected
El subproceso 0xf7c terminó con código 0 (0x0).
2022/02/20 13:42:34:195 | Debug | Binance    | Socket 1 connected to wss://stream.binance.com:9443/stream with request {"method":"SUBSCRIBE","params":["btcbusd@kline_15m"],"id":2}
2022/02/20 13:42:34:201 | Debug | Binance    | Socket 1 sending data: {"method":"SUBSCRIBE","params":["btcbusd@kline_15m"],"id":2}
2022/02/20 13:42:34:203 | Trace | Binance    | Socket 1 Adding 60 to sent buffer
'MyProgram.exe' (CoreCLR: clrhost): 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.2\System.Numerics.Vectors.dll' cargado. Se omitió la carga de símbolos. El módulo está optimizado y la opción del depurador 'Sólo mi código' está habilitada.
2022/02/20 13:42:34:216 | Trace | Binance    | Socket 1 sent 60 bytes
2022/02/20 13:42:34:466 | Trace | Binance    | Socket 1 received 22 bytes in single message
2022/02/20 13:42:34:473 | Trace | Binance    | Socket 1 received data: {"result":null,"id":2}
2022/02/20 13:42:34:497 | Trace | Binance    | Socket 1 Subscription completed
2022/02/20 13:42:34:570 | Trace | Binance    | Socket 1 received 369 bytes in single message
2022/02/20 13:42:34:574 | Trace | Binance    | Socket 1 received data: {"stream":"btcbusd@kline_15m","data":{"e":"kline","E":1645360953417,"s":"BTCBUSD","k":{"t":1645360200000,"T":1645361099999,"s":"BTCBUSD","i":"15m","f":311652312,"L":311654466,"o":"38344.60000000","c":"38365.69000000","h":"38395.60000000","l":"38266.00000000","v":"69.37355000","n":2155,"x":false,"q":"2658831.73139580","V":"24.63498000","Q":"944118.63472480","B":"0"}}}
2022/02/20 13:42:42:080 | Debug | Binance    | Disposing socket client, closing all subscriptions
2022/02/20 13:42:42:084 | Debug | Binance    | Closing all 1 subscriptions
2022/02/20 13:42:42:092 | Debug | Binance    | Socket 1 closing
2022/02/20 13:42:42:096 | Trace | Binance    | Socket 1 waiting for communication loops to finish
2022/02/20 13:42:42:098 | Debug | Binance    | Socket 1 closed
2022/02/20 13:42:42:101 | Information | Binance    | Socket 1 closed
2022/02/20 13:42:42:103 | Debug | Binance    | Socket 1 disposing
2022/02/20 13:42:42:125 | Trace | Binance    | Socket 1 disposed
2022/02/20 13:42:42:127 | Debug | Binance    | Disposing client
2022/02/20 13:42:56:483 | Debug | Binance    | Socket 1 message processing slow (14398ms), consider offloading data handling to another thread. Data from this socket may arrive late or not at all if message processing is continuously slow.
El subproceso 0x2f7c terminó con código 0 (0x0).
El subproceso 0x38c terminó con código 0 (0x0).
El subproceso 0x18ac terminó con código 0 (0x0).
El subproceso 0x97c terminó con código 0 (0x0).
El subproceso 0x4e5c terminó con código 0 (0x0).
El subproceso 0x6d8c terminó con código 0 (0x0).

I repeated the same test, and in this case no event received. It seems like a null was received to your library:

2022/02/20 13:43:51:143 | Trace | Binance    | Client configuration: LogLevel: Trace, Writers: 1, OutputOriginalData: False, Proxy: -, Base.ApiCredentials: set, AutoReconnect: True, ReconnectInterval: 00:00:15, MaxReconnectTries: , MaxResubscribeTries: 5, MaxConcurrentResubscriptionsPerSocket: 5, SocketResponseTimeout: 00:00:10, SocketNoDataTimeout: 00:00:00, SocketSubscriptionsCombineTarget: 10, CryptoExchange.Net: v5.0.0.0, Binance.Net: v8.0.0.0
2022/02/20 13:43:51:149 | Trace | Binance    |   BinanceSocketClientSpotStreams configuration: Credentials: -, BaseAddress: wss://stream.binance.com:9443/
2022/02/20 13:43:51:151 | Trace | Binance    |   BinanceSocketClientUsdFuturesStreams configuration: Credentials: -, BaseAddress: wss://fstream.binance.com/
2022/02/20 13:43:51:153 | Trace | Binance    |   BinanceSocketClientCoinFuturesStreams configuration: Credentials: -, BaseAddress: wss://dstream.binance.com/
'BazookaV1.exe' (CoreCLR: clrhost): 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.2\System.Net.WebSockets.Client.dll' cargado. Se omitió la carga de símbolos. El módulo está optimizado y la opción del depurador 'Sólo mi código' está habilitada.
'BazookaV1.exe' (CoreCLR: clrhost): 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.2\System.Net.WebSockets.dll' cargado. Se omitió la carga de símbolos. El módulo está optimizado y la opción del depurador 'Sólo mi código' está habilitada.
'BazookaV1.exe' (CoreCLR: clrhost): 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.2\System.Net.WebHeaderCollection.dll' cargado. Se omitió la carga de símbolos. El módulo está optimizado y la opción del depurador 'Sólo mi código' está habilitada.
2022/02/20 13:43:51:202 | Debug | Binance    | Socket 1 new socket created for wss://stream.binance.com:9443/stream
2022/02/20 13:43:51:211 | Debug | Binance    | Socket 1 connecting
2022/02/20 13:43:52:232 | Trace | Binance    | Socket 1 connection succeeded, starting communication
El subproceso 0x4288 terminó con código 0 (0x0).
2022/02/20 13:43:52:276 | Debug | Binance    | Socket 1 connected
El subproceso 0x52ec terminó con código 0 (0x0).
2022/02/20 13:43:52:302 | Debug | Binance    | Socket 1 connected to wss://stream.binance.com:9443/stream with request {"method":"SUBSCRIBE","params":["btcbusd@kline_15m"],"id":2}
2022/02/20 13:43:52:308 | Debug | Binance    | Socket 1 sending data: {"method":"SUBSCRIBE","params":["btcbusd@kline_15m"],"id":2}
2022/02/20 13:43:52:310 | Trace | Binance    | Socket 1 Adding 60 to sent buffer
'MyProgram.exe' (CoreCLR: clrhost): 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.2\System.Numerics.Vectors.dll' cargado. Se omitió la carga de símbolos. El módulo está optimizado y la opción del depurador 'Sólo mi código' está habilitada.
2022/02/20 13:43:52:321 | Trace | Binance    | Socket 1 sent 60 bytes
2022/02/20 13:43:52:580 | Trace | Binance    | Socket 1 received 22 bytes in single message
2022/02/20 13:43:52:593 | Trace | Binance    | Socket 1 received data: {"result":null,"id":2}
2022/02/20 13:43:52:622 | Trace | Binance    | Socket 1 Subscription completed
2022/02/20 13:43:52:627 | Debug | Binance    | Disposing socket client, closing all subscriptions
2022/02/20 13:43:52:631 | Debug | Binance    | Closing all 1 subscriptions
2022/02/20 13:43:52:639 | Debug | Binance    | Socket 1 closing
2022/02/20 13:43:52:644 | Trace | Binance    | Socket 1 waiting for communication loops to finish
2022/02/20 13:43:52:646 | Debug | Binance    | Socket 1 closed
2022/02/20 13:43:52:652 | Information | Binance    | Socket 1 closed
2022/02/20 13:43:52:654 | Debug | Binance    | Socket 1 disposing
2022/02/20 13:43:52:656 | Trace | Binance    | Socket 1 disposed
2022/02/20 13:43:52:658 | Debug | Binance    | Disposing client
El subproceso 0x1d94 terminó con código 0 (0x0).
El subproceso 0x37b4 terminó con código 0 (0x0).

**Comment by HypsyNZ** [comment]
> It worked for just one event for the first time in the first set of logs.

I don't even get one

**Comment by JKorf** [comment]
Thanks, this is useful info. So it seems like the socket is getting disposed. 

The null getting received is expected, that's how Binance responds for a successful subscription. It's weird that the client is getting disposed though. Do you keep a reference to the client? If for example you create the client and subscribe in a method and the method ends the client will get disposed, see https://jkorf.github.io/CryptoExchange.Net/FAQ.html#the-socket-client-stops-sending-updates-after-a-little-while. 

Can you confirm it's not getting disposed because there are no references to the client outside the subscription method?

**Comment by HypsyNZ** [comment]
> Can you confirm it's not getting disposed because there are no references to the client outside the subscription method?

No its not that, My implementation is based off your wpf example

reference:
https://github.com/HypsyNZ/BinanceTrader.NET/blob/70afc0b7f37109d0098ac03b0754aa06d3dade9c/ViewModels/MainViewModel.cs#L68


Open User Streams:
https://github.com/HypsyNZ/BinanceTrader.NET/blob/70afc0b7f37109d0098ac03b0754aa06d3dade9c/ViewModels/MainViewModel.cs#L805

Please keep in mind this works fine in v7 so surely you must know what you changed that would cause this behavior?

**Comment by JKorf** [comment]
I was asking specifically @arkoala since the logging seems to suggest it. It might be 2 unrelated issues. Since I don't have any logging from you its hard to say.

**Comment by arkoala** [comment]
That was exactly my issue. My fault, sorry.
The new project with different classes structure was instancing BinanceSocketClient in the same way as BinanceClient, under a using block. It is now evident socket must be alive, while client can be use and throw object.
Maybe the time I was debugging into the 'using' block was the only chance I let the object to raise an event.
Seems to be working as in 7.2.7 now for me.

My apologies interferring with SubscribeToUserDataUpdatesAsync issue.
It could be probably the same problem, or not. Extra traces from HypsyNZ would be very useful.

Wishing to finish the development, get a profitable application and donate something for this great job.

My new traces:
2022/02/20 15:14:43:521 | Debug | Binance    | Socket 1 new socket created for wss://stream.binance.com:9443/stream
2022/02/20 15:14:43:528 | Debug | Binance    | Socket 1 connecting
2022/02/20 15:14:44:557 | Trace | Binance    | Socket 1 connection succeeded, starting communication
El subproceso 0x6024 terminó con código 0 (0x0).
2022/02/20 15:14:44:580 | Debug | Binance    | Socket 1 connected
El subproceso 0x4b2c terminó con código 0 (0x0).
2022/02/20 15:14:44:602 | Debug | Binance    | Socket 1 connected to wss://stream.binance.com:9443/stream with request {"method":"SUBSCRIBE","params":["btcbusd@kline_15m"],"id":2}
2022/02/20 15:14:44:606 | Debug | Binance    | Socket 1 sending data: {"method":"SUBSCRIBE","params":["btcbusd@kline_15m"],"id":2}
2022/02/20 15:14:44:607 | Trace | Binance    | Socket 1 Adding 60 to sent buffer
'BazookaV1.exe' (CoreCLR: clrhost): 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\6.0.2\System.Numerics.Vectors.dll' cargado. Se omitió la carga de símbolos. El módulo está optimizado y la opción del depurador 'Sólo mi código' está habilitada.
2022/02/20 15:14:44:618 | Trace | Binance    | Socket 1 sent 60 bytes
2022/02/20 15:14:44:874 | Trace | Binance    | Socket 1 received 22 bytes in single message
2022/02/20 15:14:44:886 | Trace | Binance    | Socket 1 received data: {"result":null,"id":2}
2022/02/20 15:14:44:933 | Trace | Binance    | Socket 1 Subscription completed
2022/02/20 15:14:47:512 | Trace | Binance    | Socket 1 received 369 bytes in single message
2022/02/20 15:14:47:514 | Trace | Binance    | Socket 1 received data: {"stream":"btcbusd@kline_15m","data":{"e":"kline","E":1645366486332,"s":"BTCBUSD","k":{"t":1645365600000,"T":1645366499999,"s":"BTCBUSD","i":"15m","f":311667381,"L":311669306,"o":"38224.45000000","c":"38205.83000000","h":"38264.35000000","l":"38200.00000000","v":"46.18998000","n":1926,"x":false,"q":"1765727.16213980","V":"24.10280000","Q":"921395.01097300","B":"0"}}}
2022/02/20 15:14:47:542 | Trace | Binance    | Socket 1 message processed in 23ms
2022/02/20 15:14:49:730 | Trace | Binance    | Socket 1 received 369 bytes in single message
2022/02/20 15:14:49:736 | Trace | Binance    | Socket 1 received data: {"stream":"btcbusd@kline_15m","data":{"e":"kline","E":1645366488548,"s":"BTCBUSD","k":{"t":1645365600000,"T":1645366499999,"s":"BTCBUSD","i":"15m","f":311667381,"L":311669309,"o":"38224.45000000","c":"38206.94000000","h":"38264.35000000","l":"38200.00000000","v":"46.20345000","n":1929,"x":false,"q":"1766241.80926950","V":"24.11015000","Q":"921675.83162990","B":"0"}}}
2022/02/20 15:14:49:745 | Trace | Binance    | Socket 1 message processed in 1ms
2022/02/20 15:14:52:287 | Trace | Binance    | Socket 1 received 369 bytes in single message
2022/02/20 15:14:52:292 | Trace | Binance    | Socket 1 received data: {"stream":"btcbusd@kline_15m","data":{"e":"kline","E":1645366491102,"s":"BTCBUSD","k":{"t":1645365600000,"T":1645366499999,"s":"BTCBUSD","i":"15m","f":311667381,"L":311669311,"o":"38224.45000000","c":"38206.94000000","h":"38264.35000000","l":"38200.00000000","v":"46.20469000","n":1931,"x":false,"q":"1766289.18587510","V":"24.11015000","Q":"921675.83162990","B":"0"}}}
2022/02/20 15:14:52:298 | Trace | Binance    | Socket 1 message processed in 0ms
2022/02/20 15:14:55:293 | Trace | Binance    | Socket 1 received 369 bytes in single message
2022/02/20 15:14:55:296 | Trace | Binance    | Socket 1 received data: {"stream":"btcbusd@kline_15m","data":{"e":"kline","E":1645366494109,"s":"BTCBUSD","k":{"t":1645365600000,"T":1645366499999,"s":"BTCBUSD","i":"15m","f":311667381,"L":311669316,"o":"38224.45000000","c":"38206.94000000","h":"38264.35000000","l":"38200.00000000","v":"46.26214000","n":1936,"x":false,"q":"1768484.17457810","V":"24.11015000","Q":"921675.83162990","B":"0"}}}
2022/02/20 15:14:55:299 | Trace | Binance    | Socket 1 message processed in 0ms
2022/02/20 15:14:57:333 | Trace | Binance    | Socket 1 received 369 bytes in single message
2022/02/20 15:14:57:336 | Trace | Binance    | Socket 1 received data: {"stream":"btcbusd@kline_15m","data":{"e":"kline","E":1645366496153,"s":"BTCBUSD","k":{"t":1645365600000,"T":1645366499999,"s":"BTCBUSD","i":"15m","f":311667381,"L":311669317,"o":"38224.45000000","c":"38206.94000000","h":"38264.35000000","l":"38200.00000000","v":"46.26269000","n":1937,"x":false,"q":"1768505.18839510","V":"24.11015000","Q":"921675.83162990","B":"0"}}}
2022/02/20 15:14:57:340 | Trace | Binance    | Socket 1 message processed in 0ms

**Comment by JKorf** [comment]
@arkoala Thanks for the update, hope you can make something profitable 👍

**Comment by JKorf** [comment]
@HypsyNZ I've honestly tries everything I could to try and reproduce this, but I simply can't. That means there is a combination of environment variables/circumstances which causes this for issue for you which I can't recreate. 

This morning I did actually try and run you branches and after a little while I managed to get the application going, but again I did not have any issues. I tried locally recreating the same project/environment with the same reference versions and framework from scratch, but again I can't reproduce it. 

Now since you already have the code locally it should be fairly easy to find out where the code hangs or whatever else is going on. Since I don't have any logging from your end I can't help you at this moment, but if you find any issue I'd be happy to fix it.
